federation/apis/federation/v1alpha1/types.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/api/v1"
+)
+
+// ClusterSpec describes the attributes of a kubernetes cluster.
+type ClusterSpec struct {
+	// A map of client CIDR to server address.
+	// This is to help clients reach servers in the most network-efficient way possible.
+	// Clients can use the appropriate server address as per the CIDR that they match.
+	// In case of multiple matches, clients should use the longest matching CIDR.
+	ServerAddressByClientCIDRs []unversioned.ServerAddressByClientCIDR `json:"serverAddressByClientCIDRs" patchStrategy:"merge" patchMergeKey:"clientCIDR" protobuf:"bytes,1,rep,name=serverAddressByClientCIDRs"`
+	// the type (e.g. bearer token, client certificate etc) and data of the credential used to access cluster.
+	// It’s used for system routines (not behalf of users)
+	// TODO: string may not enough, https://github.com/kubernetes/kubernetes/pull/23847#discussion_r59301275
+	Credential string `json:"credential,omitempty" protobuf:"bytes,2,opt,name=credential"`
+}
+
+type ClusterConditionType string
+
+// These are valid conditions of a cluster.
+const (
+	// ClusterReady means the cluster is ready to accept workloads.
+	ClusterReady ClusterConditionType = "Ready"
+	// ClusterOffline means the cluster is temporarily down or not reachable
+	ClusterOffline ClusterConditionType = "Offline"
+)
+
+// ClusterCondition describes current state of a cluster.
+type ClusterCondition struct {
+	// Type of cluster condition, Complete or Failed.
```

Comment:
To clarify, these comments are automatically converted to user facing documentation using swagger. Users of that documentation dont see the Go struct names but only the json. Hence, we refer to json field names in these comments rather than referring to Go struct name.

Hence instead of saying "Status" in comments, we say "status". And similarly "lastProbeTime". Hope this helps.